banzai/dbs.py

```diff
@@ -356,24 +359,22 @@ def get_master_cal_record(image, calibration_type, master_selection_criteria, db
     calibration_criteria &= CalibrationImage.good_after <= image.dateobs
     calibration_criteria &= CalibrationImage.good_until >= image.dateobs
 
+    calibration_image = None
     with get_session(db_address=db_address) as db_session:
-        calibration_images = db_session.query(CalibrationImage).filter(calibration_criteria).all()
-
-    # Exit if no calibration file found
-    if len(calibration_images) == 0:
-        return None
-
-    # Find the closest date
-    date_deltas = np.abs(np.array([i.dateobs - image.dateobs for i in calibration_images]))
-    closest_calibration_image = calibration_images[np.argmin(date_deltas)]
-
-    return closest_calibration_image
-
-
-def get_master_cal(image, calibration_type, master_selection_criteria, db_address,
-                   use_only_older_calibrations=False):
-    return cal_record_to_file_info(get_master_cal_record(image, calibration_type, master_selection_criteria, db_address,
-                                                         use_only_older_calibrations=use_only_older_calibrations))
+        if prefer_same_block:
+            block_criteria = CalibrationImage.blockid == image.blockid
+            image_filter = db_session.query(CalibrationImage).filter(calibration_criteria & block_criteria)
+            calibration_image = image_filter.order_by(func.abs(CalibrationImage.dateobs - image.dateobs)).first()
+        if calibration_image is None and prefer_same_proposal:
+            proposal_criteria = CalibrationImage.proposal == image.proposal
+            image_filter = db_session.query(CalibrationImage).filter(calibration_criteria & proposal_criteria)
+            calibration_image = image_filter.order_by(func.abs(CalibrationImage.dateobs - image.dateobs)).first()
+        if check_public:
+            calibration_criteria &= CalibrationImage.public_date <= datetime.datetime.utcnow()
```

Comment:
utcnow() is deprecated  - suggested replacement is now datetime.datetime.now(datetime.UTC)